node-db/week-13/api-workshop.md

```diff
@@ -0,0 +1,540 @@
+
+[Express Body Parser](https://github.com/expressjs/body-parser#examples) makes it easier for our _endpoints_ to receive and understand different formats of data.
+
+First install the package: `npm install body-parser --save`
+
+```js
+const bodyParser = require("body-parser");
+
+app.use(bodyParser.json()); // before our routes definition
+```
+
+Now we will receive the data as `req.body`.
+
+```js
+app.post("/albums", function(req, res) {
+  console.log("POST /albums route");
+  console.log(req.body);
+});
+```
+
+> Exercise: Use Postman to `POST` this data to `/albums` endpoint.
+
+![](../assets/postman-post.png)
+
+```json
+{
+  "albumId": "13",
+  "artistName": "Beyoncé",
+  "collectionName": "B'Day (Deluxe Edition)",
+  "artworkUrl100":
+    "http://is5.mzstatic.com/image/thumb/Music/v4/6c/fc/6a/6cfc6a13-0633-f96b-9d72-cf56774beb4b/source/100x100bb.jpg",
+  "releaseDate": "2007-05-29T07:00:00Z",
+  "primaryGenreName": "Pop",
+  "url":
+    "https://www.youtube.com/embed/RQ9BWndKEgs?rel=0&amp;controls=0&amp;showinfo=0"
+}
+```
+
+> Finish the code for the route `POST /albums` to add the album data to the albums list (how to amend to an array?)
+
+# Homework: Next Steps
+
+## Delete Album
+
+Add the route `DELETE /albums/:albumId` to delete the album (that matches the passed albumId)
+
+> **Hint:** We used `app.get` to handle GET requests and `app.post` to handle POST requests. How do you think we can handle PUT and DELETE requests?
+
+## Edit Album
+
+Add the route `PUT /albums/:albumId` should update the album (that matches the passed albumId)
+
+## Allow filtering by Genre
+
+Change the route `GET /albums` to filter based on Genre (add more _mock_ data to the Albums Data with different Genre types)
+
+The endpoint should be `GET /albums?genre=pop` or `GET /albums?genre=rock` for example. The part of the url after the `?` is called query string parameter. Here is the documentation to show how you can retrieve and use it [http://expressjs.com/en/4x/api.html#req.query](http://expressjs.com/en/4x/api.html#req.query)
+
+## Back to React: Add a Song
+
+Back to React, change the React App for Beyonce App to add a form that adds a song to the list of songs. It should call `POST /songs/` in your API and pass it the song name (and other information that the User entered) and the API should save the song to the array of songs.
+
+### Integrate our React App
+
+In order to enable a client (a React app for example) that lives on a different domain to talk to our API, we need to enable Cross-origin resource sharing (CORS).
+
+> Cross-origin resource sharing (CORS) is a mechanism that allows restricted resources on a web page to be requested from another domain outside the domain from which the first resource was served.
+> [https://en.wikipedia.org/wiki/Cross-origin_resource_sharing](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing)
+
+`npm install --save cors`
```

Comment:
maybe we add this during the class .. so that they don't hit the issue at homework. It's a bit obscure thing to do.